cocos/core/assets/morph-rendering.ts

```diff
@@ -530,6 +491,100 @@ class MorphUniforms {
     }
 }
 
+/**
+ * 
+ * @param gfxDevice 
+ * @param vec4Capacity Capacity of vec4.
+ */
+function createVec4TextureFactory (gfxDevice: GFXDevice, vec4Capacity: number) {
+    const hasFeatureFloatTexture = gfxDevice.hasFeature(GFXFeature.TEXTURE_FLOAT);
+
+    let pixelRequired: number;
+    let pixelFormat: PixelFormat;
+    let pixelBytes: number;
+    let updateViewConstructor: typeof Float32Array | typeof Uint8Array;
+    if (hasFeatureFloatTexture) {
+        pixelRequired = vec4Capacity;
+        pixelBytes = 16;
+        pixelFormat = Texture2D.PixelFormat.RGBA32F;
+        updateViewConstructor = Float32Array;
+    } else {
+        pixelRequired = 4 * vec4Capacity;
+        pixelBytes = 4;
+        pixelFormat = Texture2D.PixelFormat.RGBA8888;
+        updateViewConstructor = Uint8Array;
+    }
+
+    const { width, height } = bestSizeToHavePixels(pixelRequired);
+    assertIsTrue(width * height > pixelRequired);
+
+    return {
+        width,
+        height,
+        create: () => {
+            const arrayBuffer = new ArrayBuffer(width * height * pixelBytes);
+            const valueView = new Float32Array(arrayBuffer);
+            const updateView = updateViewConstructor === Float32Array ? valueView : new updateViewConstructor(arrayBuffer);
+            const image = new ImageAsset({
+                width,
+                height,
+                _data: updateView,
+                _compressed: false,
+                format: pixelFormat,
+            });
+            const textureAsset = new Texture2D();
+            textureAsset.setFilters(Texture2D.Filter.NEAREST, Texture2D.Filter.NEAREST);
+            textureAsset.setMipFilter(Texture2D.Filter.NONE);
+            textureAsset.setWrapMode(Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE);
+            textureAsset.image = image;
+            if (!textureAsset.getGFXTexture()) {
+                warn(`Unexpected: failed to create morph texture?`);
+            }
+            const sampler = samplerLib.getSampler(gfxDevice, textureAsset.getSamplerHash());
+            return {
+                /**
+                 * Gets the GFX texture.
+                 */
+                get texture () {
+                    return textureAsset.getGFXTexture()!;
+                },
+
+                /**
+                 * Gets the GFX sampler.
+                 */
+                get sampler () {
+                    return sampler;
+                },
+
+                /**
+                 * Value view.
+                 */
+                get valueView () {
+                    return valueView;
+                },
+
+                /**
+                 * Destroy the texture. Release its GPU resources.
+                 */
+                destroy () {
+                    textureAsset.destroy();
+                    // TODO: Should we free sampler?
+                    // this._sampler.destroy();
```

Comment:
Please remove the TODO and code

Reply:
https://github.com/cocos-creator/engine/pull/6937/commits/9877e50da83a644e8b37f233f22a4bfeb9a4c212